### ISSUE TITLE
Issues with xpath

### ISSUE THREAD

**Issue by oli-couvreur** [google/gnxi#63]
Hello,

I'm facing two issues with the xpath parsing both resulting in a "xpath component parse error".

It seems that '-' are not supported in the xpath when used for a key like in:
--xpath '/state/chassis[chassis-number=1]/name/'
It results in a "__main__.XpathError: xpath component parse error: chassis[chassis-number=1]".

It seems that '/' are stripped from the xpath, even when used as a key value like in:
--xpath '/state/port[port_id=1/1/1]/statistics/out-packets'
It results in a "__main__.XpathError: xpath component parse error: port[port_id=1"

(I cross-checked the xpath syntax with gnmi_get, which does not give errors on the xpaths described above)

Any ideas ?

Thanks,
Oli

**Comment by tejaskumark** [comment]
@oli-couvreur Please replace in your script _RE_PATH_COMPONENT with below one,

_RE_PATH_COMPONENT = re.compile(r'''
^
(?P<pname>[^[]+)  # gNMI path name
(\[(?P<key>[a-zA-Z0-9\-]+)   # gNMI path key
=
(?P<value>.*)    # gNMI path value
\])?$
''', re.VERBOSE)

Hopefully it will work.
Thanks.

**Comment by oli-couvreur** [comment]
Hi, thank you for the fix on the hyphen '-'.
It's working fine.

I open another issue for the '/' in the (key, value), to make it easier to track.

**Comment by batmancn** [comment]
> Hi, thank you for the fix on the hyphen '-'.
> It's working fine.
> 
> I open another issue for the '/' in the (key, value), to make it easier to track.

I have same problem, `/xxx/xxx[xxx=1/2/3]/xxx`, need to be parse correctly

**Comment by tejaskumark** [comment]
@batmancn @oli-couvreur 
I hope you are using this regex in your code as below, if not please change it as below,

_RE_PATH_COMPONENT = re.compile(r'''
^
(?P<pname>[^[]+)  # gNMI path name
(\[(?P<key>[a-zA-Z0-9\-]+)   # gNMI path key
=
(?P<value>.*)    # gNMI path value
\])?$
''', re.VERBOSE)

If it is not working can you please share output of request when you use through cli.

**Comment by batmancn** [comment]
> @batmancn @oli-couvreur
> I hope you are using this regex in your code as below, if not please change it as below,
> 
> # _RE_PATH_COMPONENT = re.compile(r'''
> ^
> (?P[^[]+) # gNMI path name
> ([(?P[a-zA-Z0-9-]+) # gNMI path key
> (?P.*) # gNMI path value
> ])?$
> ''', re.VERBOSE)
> 
> If it is not working can you please share output of request when you use through cli.

It's NOT work

I use this
```
_RE_PATH_COMPONENT = re.compile(r'''
^
(?P[^[]+) # gNMI path name
([(?P[a-zA-Z0-9-]+) # gNMI path key
(?P.*) # gNMI path value
])?$
''', re.VERBOSE)
```

I test by this
```
gyw@sonic107:~/go/src/github.com/google/gnxi/gnmi_cli_py$ python py_gnmicli.py  -n  -m set-delete -t 172.18.8.241 -p 8080 -x "/interfaces/interface[name=Ethernet16]/subinterfaces/subinterface/ipv4/addresses/address/config/ip[address=1.1.1.1/24]" -xt "MTNOS"
Traceback (most recent call last):
  File "py_gnmicli.py", line 58, in <module>
    ''', re.VERBOSE)
  File "/usr/lib/python2.7/re.py", line 194, in compile
    return _compile(pattern, flags)
  File "/usr/lib/python2.7/re.py", line 251, in _compile
    raise error, v # invalid expression
sre_constants.error: unknown specifier: ?P[
```

**Comment by tejaskumark** [comment]
@batmancn I think your your file py_gnmicli.py having some issue, below is output from xpath you have used on my setup, 

Request you to use file available at repo - https://github.com/tejaskumark/gnxi/tree/patch-3/gnmi_cli_py 
 ------------------------- 
 elem {
  name: "interfaces"
}
elem {
  name: "interface"
  key {
    key: "name"
    value: "Ethernet16"
  }
}
elem {
  name: "subinterfaces"
}
elem {
  name: "subinterface"
}
elem {
  name: "ipv4"
}
elem {
  name: "addresses"
}
elem {
  name: "address"
}
elem {
  name: "config"
}
elem {
  name: "ip"
  key {
    key: "address"
    value: "1.1.1.1/24"
  }
}

**Comment by mike-albano** [comment]
Both of these fixed, in PR #223 and PR #240
